### PR TITLE
Added edge input feature to fast_line_detector

### PR DESCRIPTION
**Pull request by tsukada-cs**
Fixed issue #2667 in opencv/opencv_contrib

Added the `is_edge` option to the `detect` method in `fast_line_detector`.

The Fast Line Detector used to use the canny method for edge detection internally, but `is_edge=true` makes it possible to input the edge image directly. Parameters for the canny method are ignored. The default value is `false`, so it won't break anyone's code.

This feature will provide users with additional flexibility and versatility in FastLineDetector.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/ximgproc/include/opencv2/ximgproc/fast_line_detector.hpp [hunk 1]
```diff
@@ -40,8 +40,10 @@ class CV_EXPORTS_W FastLineDetector : public Algorithm
       and ending point of a line.  Where Vec4f is (x1, y1, x2, y2), point
       1 is the start, point 2 - end. Returned lines are directed so that the
       brighter side is on their left.
+      @param is_edge If true, image will be considerd as edge and negrect the canny parameters.
       */
-    CV_WRAP virtual void detect(InputArray _image, OutputArray _lines) = 0;
+    CV_WRAP virtual void detect(InputArray _image, OutputArray _lines,
+            bool is_edge = false) = 0;
```
Comment (alalek) [inline modules/ximgproc/include/opencv2/ximgproc/fast_line_detector.hpp:46]: `detect()` is not a right place for algorithm parameters tuning. See `createFastLineDetector()`

Can we check some canny parameters to handle that without touching API?
For example, `canny_aperture_size == 0`

**Comment by tsukada-cs** [comment]
@alalek 
Thanks for the comment.
I tried changing the internal assert statement and then `canny_aperture_size = 0`, `canny_th1=0`, `canny_th2=0`, etc., but it didn't work.
So I removed the `is_edge` option of `detect()` and added the `input_edge` option to `createFastLineDetector()`.

**Comment by alalek** [comment]
> but it didn't work as expected

Try to split this assertion check first.

BTW, There is no sense to define `canny_*` parameters if we don't use that.

**Comment by tsukada-cs** [comment]
@alalek 

> Try to split this assertion check first.

I failed to state the situation correctly.
I split the assertion check and tried `canny_appature_size=0`, but the canny method was applied as a result.
Also, the documentation says that the `ksize` of `Sobel()` must be 1, 3, 5 or 7, so we are not supposed to enter `canny_appature_size = 0` (`ksize=0`).

> BTW, There is no sense to define canny_* parameters if we don't use that.

I agree.
Should we change the implementation?

**Comment by alalek** [comment]
> and tried `canny_appature_size=0`, but the canny method was applied as a result

You should add this check before `Canny()` call to bypass it.

**Comment by tsukada-cs** [comment]
@alalek 

Thanks for the kind comments.
I changed the implementation to use the input image as an edge image when `canny_aperture_size=0`.

File: modules/ximgproc/src/fast_line_detector.cpp [hunk 2]
```diff
@@ -544,7 +545,8 @@ void FastLineDetectorImpl::lineDetection(const Mat& src, std::vector<SEGMENT>& s
     std::vector<Point2i> points;
     std::vector<SEGMENT> segments, segments_tmp;
     Mat canny;
-    Canny(src, canny, canny_th1, canny_th2, canny_aperture_size);
+    if (canny_aperture_size == 0) canny = src;
+    else Canny(src, canny, canny_th1, canny_th2, canny_aperture_size);
```
Comment (alalek) [inline modules/ximgproc/src/fast_line_detector.cpp:549]: Consider using dedicated lines for if/else bodies.
It is necessary to see real picture from "code coverage" line-based reports ([example of weekly report](http://pullrequest.opencv.org/buildbot/export/opencv_releases/master-contrib_coverage-lin64-debug/20201017-080737_10545/coverage_html/index.html)).

File: modules/ximgproc/src/fast_line_detector.cpp [hunk 3]
```diff
@@ -29,10 +29,11 @@ class FastLineDetectorImpl : public FastLineDetector
...
          * @param _canny_th2           50         - Second threshold for
          *        _                                 hysteresis procedure in Canny()
-         * @param _canny_aperture_size 3          - Aperturesize for the sobel
-         *        _                                 operator in Canny()
+         * @param _canny_aperture_size 3          - Aperturesize for the sobel operator in Canny().
+         *                                          If zero, Canny() is not applied and the input
+         *                                          image is taken as an edge image.
          * @param _do_merge            false      - If true, incremental merging of segments
-                                                   will be perfomred
+         *                                          will be perfomred
```
Comment (sturkmen72) [inline modules/ximgproc/src/fast_line_detector.cpp:36]: will be performed

**Comment by tsukada-cs** [comment]
@alalek 

> Consider using dedicated lines for if/else bodies.
> It is necessary to see real picture from "code coverage" line-based reports (example of weekly report).

I tried, but is this what you mean?

File: modules/ximgproc/src/fast_line_detector.cpp (hunk 3 above)
Comment (tsukada-cs) [reply modules/ximgproc/src/fast_line_detector.cpp]: Thanks for pointing that out. Fixed it.

**Review by alalek — APPROVED** [review]
Well done! Thank you :+1:

**Comment by tsukada-cs** [comment]
Thank you for accepting it.  Thanks for all the kind comments.

**Pull request merged**